Improve SAML error handling by adding metadata

piotrsulkowski-elastic opened the pull request:
in case when there is a mismatch between the expected request ID (that comes from a cookie) and the actual in-response-to field in the SAML content, return more specific error to allow for better error handling at the client

closes #128179

piotrsulkowski-elastic committed:
Improve SAML error handling by adding metadata

in case when there is a mismatch between the expected request ID
(that comes from a cookie) and the actual in-response-to field in the
SAML content, return more specific error to allow for better error
handling at the client

closes #128179

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

jfreden reviewed (commented):
Great job with this PR! I have a couple of comments but the overall flow LGTM! 

I think you're missing a changelog since this is an `>enhancement`, maybe the autogenerated one was overwritten during a force-push? 

It would be good with a more descriptive title too: "Improve SAML Error Handling by Including additional Metadata" or something like that?

piotrsulkowski-elastic committed:
code review improvements

elasticsearchmachine commented:
Hi @piotrsulkowski-elastic, I've created a changelog YAML for you.

piotrsulkowski-elastic committed:
Update docs/changelog/137598.yaml

piotrsulkowski-elastic committed:
fix NullPointerException in SamlToken

piotrsulkowski-elastic committed:
create SamlAuthenticationException

create a separate exception class for SAML-related errors (which
previously was handled using metadata). Also, allow this exception
to support TERMINATION/CONTINUE flag, which was previously not possible,
and CONTINUE was implicitly assumed (which resulted in wiping
all the exception metadata from the resulting transport and
rest responses).

piotrsulkowski-elastic committed:
integration test for InResponseTo error

Creates a new integration test, which checks the handling of errors in
various cases of in_response_to header vs saml token metadata mismatch.
Extracts common test base for saml IT tests from the existing
SamlServiceProviderMetadataIT

piotrsulkowski-elastic commented:
I was making end-to-end testing and discovered that it wasn't working the way I expected due to 2 factors:

- there was a null pointer exception thrown when parsing mismatched in-response-to headers - that got an easy fix (but took me a while to find it, because NPE was caught in one of the listeners and transformed into generic unauthorized exception, which looked like intentional flow)

- it wasn't working still, because saml exceptions are gathered in a loop, iterating over all saml realms. Once all exceptions are collected, they are written to log, but the final exception is generic with all metadata excluded. I propose here a solution where the saml exception can be either terminating, or not. The terminating exceptions break the iteration over realms, and are propagated straight away to the transport and then rest responses with all metadata. The non-terminating ones behave just as they used to be before. That gives us more control and allows to capture both use cases, but unfortunately required some refactoring of code, which I appreciate may be discussible - please let me know your opinions on this approach.

Lastly, I added an integration test to check this behaviour (I found it appropriate given two non-trivial already existing problems).

jfreden reviewed (commented):
Ah yes, you're right, it's all supplied in the request so the concern of this validation causing early unexpected terminations isn't valid. Sorry for the confusion! 

I liked Tim's suggestion of:
>Can't we just say that the enhanced error message is only returned if a specific realm is targeted by the request? 

So we only terminate the auth with the enhanced error message if a realm is supplied? That way we don't risk breaking any bwc, even though that seems highly unlikely. Maybe if the customer has a custom realm? 

Code looks good, left some suggestions.

piotrsulkowski-elastic committed:
Revert "create SamlAuthenticationException"

This reverts commit 61956741d7f951d94cc5d42a46e080a2912a12a5.

piotrsulkowski-elastic committed:
refactoring after code review

piotrsulkowski-elastic commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlUtils.java:
```diff
@@ -96,32 +96,28 @@ static void initialize(Logger logger) throws PrivilegedActionException {
-    public static ElasticsearchSecurityException samlException(String msg, Object... args) {
-        final ElasticsearchSecurityException exception = new ElasticsearchSecurityException(msg, args);
-        exception.addMetadata(SAML_EXCEPTION_KEY);
+    static SamlAuthenticationException samlUnsolicitedInResponseToException(
```
same as above

jfreden reviewed (commented):
I think the code looks great! Good job getting all the requirements in order, the result turned out really clean. Just have a discussion point.

jfreden reviewed (approved):
LGTM!

piotrsulkowski-elastic committed:
refactoring after code review
